Python nexus client cleanup

jmportug opened the pull request.

jmportug committed:
Set up test cases from old client to run with the new client
Not all tests are currently passing

jmportug committed:
Most test cases are now working
* group membership management tests are still not passing
* there are existing policies for new groups
	- this causes 200 status code responses instead of 201 status cod responses
	- this also causes the returned content to be different from the input policies
	- lines 236-253 of go_rest_client.py

jmportug committed:
Mostly cleaned up the code based on the last code review
Added a few notes on possible problems with email validation

jmportug committed:
Merged code with similar functionality and Deleted unused files

Merged Client and GlobusOnlineRestClient
- the new merged class is called MergedClient (at least for now)
- still need to refactor the MergedClient

Merged the Client and GlobusOnlineRestClient test cases
- the new merged test cases are in merged_client_tests.py in lib/merged_tests
- this includes the tests form the transfered go_rest_cleint_tests
- fomated the test cases to be a little more uniform

Deleted the vendor directory

Deleted the 'new' tests folder containing the transfed go_rest_client_tests
- no longer need this folder because the tests have been merged

Moved the test_utils folder to lib/test_utils
- needed to move because the merged_tests require test_utils
- also because the original tests have been deleted

jmportug committed:
Got the sample to work with the MergedClient
* Changed client in use_client.py to use two MergedClients
        - there is now an alias_client and user_client
* Changed tests so they consistently use client_id instead of clientid
        - all tests are now passing except for email, policies and group list tests
* Cleaned up policy and GoAuth tests a little
* Changed MergedClient constructor so it uses a config file for all values
* Added goauth prefix to goauth functions
* Removed go_host form merged_client
        - all calls to self.go_host now use self.server from the config file
* Removed client.py and go_rest_client.py in lib/nexus
* Removed lib/tests directory
        - all tests are in lib/merged_tests now

mattias-lidman committed:
Added a bunch of dependecies.

jmportug committed:
Refactored goauth methods and other related methods

* set the goauth_token to None in the logout() method
	- might have solved the logout/logging in bug
* created temporary oauth and goauth variable in constructor
	- will help prevent bugs because methods are pass-by-reference
* changed the constructor login attempt to use the client from the config
* changed the appropriate goauth methods to use _issue_rest_request
	- ssl is no longer verified (probably will want to reimplement later)
* changed _issue_rest_request to be compatible with 'text/html' returned content-type
* renamed merged_tests directory to tests
* renamed MergedClient to GlobusOnlineRestClient
* added another sample called sample2
	- meant as a sanity check for goauth login

jmportug committed:
A lot of cleanup
* added a delete_group() method
* added a script to clean up the left over test groups from old test runs
	- can be adapted to clean any set of groups
* cleaned up tests further
* split the tests into separate goauth, user and group test files

jmportug committed:
Added post, get and delete methods for user rsa keys
Also added simple testcases for the new rsa methods

jmportug committed:
Added rsa key posting to the sample

changed goauth_get_user_usering_access_token to not use _issue_rest_request
	- the sample had stopped working

jmportug committed:
sample now runs on production
removed cleanup_groups.py and config.py

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -23,23 +24,68 @@
+import requests
+import rsa
+
+log = logging.getLogger()
```
This will get you the root logger. You might want to do `logging.getLogger(__name__)` to get a logger named for this module. http://docs.python.org/2/library/logging.html#logging.getLogger

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -23,23 +24,68 @@
+log = logging.getLogger()
+
+# using the 'new-style class' format to ensure
+# compatibility with possible NexusClient inherencies
```
typo? "inherencies'?

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -23,23 +24,68 @@
+        elif config is not None:
+            self.config = config
+        else:
+            raise AttributeError("No configuration was specified")
```
This is super pedantic and minor, but AttributeError is for failures dealing with attributes, not arguments passed in. This should probably be a ValueError (or maybe a TypeError): http://docs.python.org/2/library/exceptions.html#exceptions.ValueError

Discussion: http://stackoverflow.com/questions/256222/which-exception-should-i-raise-on-bad-illegal-argument-combinations-in-python

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -286,20 +336,19 @@ def get_user(self, username, fields=None, custom_fields=None, use_session_cookie
-        url = '/users/' + username + '?' + urllib.urlencode(query_params)
         return self._issue_rest_request(url, use_session_cookies=use_session_cookies)
-
+    
```
Random whitespace characters?

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -374,7 +423,7 @@ def username_password_login(self, username, password=None):
                 "Could not retrieve user secret.")
         self.oauth_secret = secret_content['secret']
-        self.current_user = username
+        self.client = username
```
As discussed, maybe some comments or docstring to make it clearer what `client` is.

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -416,6 +466,27 @@ def post_email_validation(self, validation_code):
+        elif rsa_key is not None:
+             key = rsa_key
+        else:
+            raise AttributeError("No rsa key was specified")
```
Same as above--ValueError.

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -416,6 +466,27 @@ def post_email_validation(self, validation_code):
+        return self._issue_rest_request(path, http_method='POST', params=params)
+
+    def get_rsa_key_list(self):
+        path = '/users/'+self.client+'/credentials'
```
Incredibly super pedantic, but:

`'/'.join(('users', self.client, 'credentials'))` is faster/more efficient than using the +. You might also consider `os.path.join` or `urlparse.urljoin`.

References:
http://stackoverflow.com/questions/1793261/how-to-join-components-of-a-path-when-you-are-constructing-a-url-in-python
http://paolobernardi.wordpress.com/2012/11/06/python-string-concatenation-vs-list-join/
http://www.skymind.com/~ocrow/python_string/

bd4 replied:
I often use the string format operator for things like this: `'/users/%s/credentials' % self.client`. I think it's more of a style question than a performance question, since only a small constant number of strings are involved.

bjmc replied:
Agreed. In this particular case, I'm sure the performance impact is totally negligible. It was more of an FYI.

On the subject of style, I do like string formatting here, but % is deprecated. Go with `'/users/{0}/credentials'.format(self.client)` to be forward-compatible for Python 3.

bd4 replied:
I'm sad they deprecated %, .format is longer and doesn't stand out as well. Why you be so anti operator Python? Time to switch to Perl 6...

bjmc replied:
Haha. I'm glad they deprecated it. It makes no sense that % means "modulo" and also "string format" ... those things have nothing in common.

bd4 replied:
I have no problem with the same character be used for two unrelated
operators. If they were related then it might be confusing. Also printf
family has used % to mark format specifiers, so the % has a nice
association with formatting and interpolation.

Back to more relevant discussion, what version of Python are we
targeting? I think we should at least support 2.6, which does have
str.format. This is what the transfer-api requires - I've had a few
users with older Python in restrictive environments, but they found
workarounds eventually and making it work with 2.4/5 never became a
priority. In general I like to support older versions when it's not too
hard. Sometimes it's better to use the old deprecated stuff,
depending on user requirements.

jbryan replied:
I know I'm not on this team any more .... but I still get these messages and just had to respond.  While those articles do address the asymptotic behavior of various string concatenation options, in practice, constants matter:

In [11]: client = 'blah'

In [12]: %timeit '/users/' + client + '/credentials/ssh2'
10000000 loops, best of 3: 136 ns per loop

In [13]: %timeit '/'.join(('/users', client, 'credentials/ssh2'))
1000000 loops, best of 3: 197 ns per loop

In [14]: %timeit '/users/%s/credentials/ssh2' % client
10000000 loops, best of 3: 149 ns per loop

... so in this case, the function call and implicit tuple creation in the join call overshadow the cost of the additional memory allocations.   Anyway, hope all is going well!  Also, @bd4, I too will miss the % operator.

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -453,26 +524,25 @@ def _construct_group_filter(self, my_roles, my_statuses):
-        url = self.go_host + path
-        headers = {}
+        url = 'https://' + self.server + path
+        if headers is None: headers = {}
```
It's a little weird having the colon here for a one-liner (does it work?). You could do:

`headers = headers or {}`

or

 `headers = headers if headers is not None else {}`

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -480,11 +550,12 @@ def _issue_rest_request(self, path, http_method='GET', content_type='application
             self.session_cookies = response['set-cookie']
         if 'content-type' in response and 'application/json' in response['content-type'] and content != '':
             return response, json.loads(content)
+        elif response['status'][0] != '4' and 'content-type' in response and 'text/html; charset=UTF-8' in response['content-type'] and content != '':
```
Maybe put a #TODO or # FIXME comment on here so somebody will know to take this out later. Seems a bit like a hack where maybe we should be returning `application/json`

Also, consider:

`elif int(response['status']) not in range(400, 500)`

or

`elif not response['status'].startswith('4')`

bd4 replied:
Looking for exactly 'text/html; charset=UTF-8' is fragile. I guess if it's a temporary hack anyway leave it, but in general content-type headers should be parsed in an RFC complient way.

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -533,7 +604,112 @@ def _put_membership_status_wrapper(self, gid, username, new_status, expected_cur
+
+    def goauth_generate_request_url(self, username=None):
+        """
+        In order for the user to authorize the client to access his data, he
```
`he` -> `he or she` / `they`

:tongue: :womens:

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -533,7 +604,112 @@ def _put_membership_status_wrapper(self, gid, username, new_status, expected_cur
+        url_parts = ('https', self.server, '/goauth/token', None, None)
+        result = token_utils.request_access_token(self.client,
+                self.client_secret, code, urlparse.urlunsplit(url_parts))
+        return (
```
Namedtuples are awesome for uses like this: http://docs.python.org/2/library/collections.html#collections.namedtuple

bjmc commented on lib/nexus/go_rest_client.py:
```diff
@@ -533,7 +604,112 @@ def _put_membership_status_wrapper(self, gid, username, new_status, expected_cur
+                username,
+                query=query_params,
+                password=password)
+        url = path + '?' + query_params
```
Can you use urlunsplit (or urlunparse?) like you did above, instead of building this URL yourself?

bjmc commented on lib/test_utils/smtp_server.py:
```diff
@@ -0,0 +1,81 @@
+import sys, asyncore, threading, socket, smtpd, time, StringIO
+
+class SmtpMailsinkServer(smtpd.SMTPServer):
```
What is this for? You say it doesn't actually work anymore?

bjmc commented on lib/test_utils/smtp_server.py:
```diff
@@ -0,0 +1,81 @@
+    TIME_TO_WAIT_BETWEEN_CHECKS_TO_STOP_SERVING = 0.001
+
+    def __init__( self, host = "localhost", port = 8025, mailboxFile = None, threadName = None ):   
+        self.throwExceptionIfAddressIsInUse( host, port )
```
somebodyHasBeenWritingTooMuchJava

bjmc commented on lib/tests/goauth_tests.py:
```diff
@@ -1,32 +1,31 @@
                     },
-                "server": "graph.api.globusonline.org",
+                "server": "graph.api.go.sandbox.globuscs.info",
                 "client": "I am not a client",
```
_ceci n'est pas une pipe_

bjmc commented on lib/tests/groups_tests.py:
```diff
@@ -0,0 +1,376 @@
+from nexus.go_rest_client import UnexpectedRestResponseError
+from test_utils.smtp_server import SmtpMailsink
+
+class TestMergedClient(unittest.TestCase):
```
I guess this goes to our earlier conversation about getting these tests running on Jenkins, but maybe investigate what it would take to mock out `requests` and return stored responses so they could be run safely without touching a database.

bd4 replied:
On Wed, 24 Jul 2013 12:46:35 -0700
bjmc notifications@github.com wrote:

> > +import unittest
> > +import json
> > +import httplib2
> > +import StringIO
> > +import email
> > +import mailbox
> > +import re
> > +import random
> > +import string
> > +
> > +from nose.plugins.attrib import attr
> > +from nexus.go_rest_client import GlobusOnlineRestClient
> > +from nexus.go_rest_client import UnexpectedRestResponseError
> > +from test_utils.smtp_server import SmtpMailsink
> > +
> > +class TestMergedClient(unittest.TestCase):
> 
> I guess this goes to our earlier conversation about getting these
> tests running on Jenkins, but maybe investigate what it would take to
> mock out `requests` and return stored responses so they could be run
> safely without touching a database.
> Would you have a script for updating the stored responses from a
> current running server? I'm skeptical that tests involving stored
> responses could be useful without that.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/globusonline/python-nexus-client/pull/1/files#r5381976

jmportug committed:
Cleaned up code according to comments on pull request

jmportug committed:
General cleanup
added test_config_file.py
	- used by the tests as a universal config_file
removed sample2
only use client_secret for password checks now
	- correct client_secret will be lost for oauth and goauth logins

jmportug committed:
General cleanup
